File: code/go/day10/day10.go
package day10

func HasPath(matrix string, rows, cols int, str string) bool {
	// 非法输入返回false
	if matrix == "" || rows < 1 || cols < 1 || str == "" {
		return false
	}

	visited := make([]bool, rows * cols)
	
	pathLength := 0

	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			// 起始点
			if hasPathCore(matrix, rows, cols, row, col, pathLength, str, visited) {
				return true
			}
		}
	}
	return false
}

func hasPathCore(matrix string, rows, cols, row, col, pathLength int, str string, visited []bool) bool {
	hasPath := false

	// 长度等于查找字符长度时即推出
	if pathLength == len(str) {
		return true
	}

	if row >= 0 && row < rows && col >= 0 && col < cols && matrix[row*cols+col] == str[pathLength] && !visited[row*cols+col] {
		pathLength++
		visited[row*cols+col] = true

		// 向上下左右前进一步，匹配成功则为true
		hasPath = hasPathCore(matrix, rows, cols, row, col-1, pathLength, str, visited) || hasPathCore(matrix, rows, cols, row-1, col, pathLength, str, visited) || hasPathCore(matrix, rows, cols, row, col+1, pathLength, str, visited) || hasPathCore(matrix, rows, cols, row+1, col, pathLength, str, visited)

		// 向前一步没有匹配的路径，失败则退回
		if !hasPath {
			pathLength--
			visited[row * cols + col] = false
		}
	}

	return hasPath
}

File: code/python/day10/day10.py
"""
请设计一个函数，用来判断在一个矩阵中是否存在一条包含某字符串所有字符路径。
路径可以从矩阵任意一格开始，每一步可以在矩阵中向左、右、上下移动一格。
如果一条路径经过了矩阵的某一格，那么该路径不能再次进入该格子。
a b t g
c f c s 
j d e h 
包含一条字符串"bfce"的路径。但矩阵中不包含"abfb"的路径。
"""
import unittest


def has_path(matrix, cols, rows, string):
    visited = [False] * (rows*cols)
    path_len = 0
    for row in range(rows):
        for col in range(cols):
            # 从任意点出发
            if has_path_core(matrix, visited, cols, col, rows, row, string, path_len):
                return True

    return False

def has_path_core(matrix, visited, cols, col, rows, row, string, path_len):
    if path_len == len(string):
        return True

    has_path = False
    if col < 0 or row < 0 or col >= cols or row >= rows:
        return False

    if not visited[cols*row + col] and matrix[row][col] == string[path_len]:
        visited[cols*row + col] = True
        path_len += 1
        has_path = has_path_core(matrix, visited, cols, col+1, rows, row, string, path_len) or \
                   has_path_core(matrix, visited, cols, col, rows, row+1, string, path_len) or \
                   has_path_core(matrix, visited, cols, col-1, rows, row, string, path_len) or \
                   has_path_core(matrix, visited, cols, col, rows, row-1, string, path_len)
        if not has_path:
            # 不符合即回溯
            visited[cols*row + col] = False
            path_len -= 1

    return has_path


class HasPathTestCase(unittest.TestCase):
    def test_1(self):
        matrix = [['a', 'b', 't', 'g'],
                  ['c', 'f', 'c', 's'],
                  ['j', 'd', 'e', 'h']]

        self.assertEqual(has_path(matrix, 4, 3, 'bfce'), True)
        self.assertEqual(has_path(matrix, 4, 3, 'abfb'), False)

    def test_2(self):
        matrix = [['a', 'a', 'a', 'a'],
                  ['a', 'a', 'a', 'a'],
                  ['a', 'a', 'a', 'a']]

        self.assertEqual(has_path(matrix, 4, 3, 'aaaaaaaaaaaaaaaa'), False)


if __name__ == '__main__':
    unittest.main()
